spring-ai-core/src/main/java/org/springframework/ai/chat/client/advisor/RetrievalAugmentationAdvisor.java

```diff
@@ -135,17 +111,47 @@ private AdvisedRequest before(AdvisedRequest request) {
 			transformedQuery = queryTransformer.apply(transformedQuery);
 		}
 
-		// 2. Retrieve similar documents for the original query.
-		List<Document> documents = this.documentRetriever.retrieve(transformedQuery);
+		// 2. Expand query into one or multiple queries.
+		List<Query> expandedQueries = queryExpander != null ? queryExpander.expand(transformedQuery)
+				: List.of(transformedQuery);
+
+		// 3. Get similar documents for each query.
+		Map<Query, List<List<Document>>> documentsForQuery = expandedQueries.stream()
+			.map(query -> CompletableFuture.supplyAsync(() -> getDocumentsForQuery(query), taskExecutor))
```

Comment:
Query expansion and/or routing can result in parallel executions that is supported by a `TaskExecutor`, following the common practice in Spring. By default, a `ThreadPoolTaskExecutor` is used, but it's possible to customise it. For example, you can pass the auto-configured `TaskExecutor` from Spring Boot or another implementation, such as `VirtualThreadTaskExecutor`. I'll describe this in detail in the separate PR where I'm writing the documentation.

Reply:
I think is a good starting point that is generally useful.  Users can implement their own RAG Advisor to customize how they do scatter gather based on other frameworks and patterns.